Add 2D Right-hand-side support for cuDSS backend `linalg.solve`

zitongzhan opened the pull request:
Support for solving sparse CSR linear systems with 2D right-hand side (RHS) tensors when linear system is sparse.
Fix https://github.com/pytorch/pytorch/pull/129856#issuecomment-2907842366

- `_apply_sparse_csr_linear_solve` in `SparseCsrTensorMath.cu` accepts both 1D and 2D RHS tensors, with additional stride and size checks for 2D cases. The function now computes leading dimensions and number of RHS columns (`nrhs`, `b_ld`, `x_ld`) appropriately for batched solves. 
- `_sparse_csr_linear_solve` converts RHS to column-major layout as required by cuDSS, and copies results back to the expected output shape.

**Testing**
- Updated `test_sparse_csr.py` to remove the restriction against 2D RHS.

Minimal code to reproduce:
```
import torch
torch.manual_seed(0)

spd = torch.rand(8, 5, device="cuda", dtype=torch.float32)
A = (spd.T @ spd).to_sparse_csr()

# 2D RHS (n, k) is supported by cuDSS-backed sparse CSR solve.
B = torch.rand(A.size(0), 3, device="cuda", dtype=torch.float32)

X_sparse = torch.linalg.solve(A, B)
X_dense = torch.linalg.solve(A.to_dense(), B)

max_err = (X_sparse - X_dense).abs().max().item()
print("max_abs_err:", max_err)
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @Lucaskabela

zitongzhan committed:
[incomplete] Enhance sparse CSR linear solve to support 2D tensors as RHS

zitongzhan committed:
avoid overly strict check on fake 1d tensor

zitongzhan committed:
Fix cuDSS CUDSS_STATUS_INVALID_VALUE error for sparse CSR solve with 2D RHS

Fixed a bug in the cuDSS-backed sparse CSR linear solver where solving
with a 2D right-hand side tensor of shape (n, 1) would fail with:
`cudss error: CUDSS_STATUS_INVALID_VALUE, when calling cudssExecute()`

The issue occurred because for column-major layout, cuDSS requires the
leading dimension (ld) to be >= nrows. When the RHS has shape (n, 1)
and is converted to column-major with stride (1, 1), the stride(1)=1
was used as the leading dimension, but cuDSS requires ld >= n.

The fix ensures that for column-major matrices, the leading dimension
is always max(size(0), stride(1)) to satisfy cuDSS requirements.

Before the fix:
- 1D RHS (n,): ✓ worked
- 2D RHS (n, 1): ✗ failed with CUDSS_STATUS_INVALID_VALUE
- 2D RHS (n, k) where k > 1: ✓ worked

After the fix:
- All cases work correctly

Reproduction script (test_reproduce_bug.py):
```python
import torch

if not torch.cuda.is_available():
    print("SKIP: CUDA is not available.")
    exit(0)

torch.manual_seed(0)

# Build a small SPD matrix
spd = torch.rand(8, 5, device="cuda", dtype=torch.float32)
A = (spd.T @ spd).to_sparse_csr()

# Test case 1: 1D RHS (should work)
print("Test 1: 1D RHS (5,)")
b1 = torch.rand(A.size(0), device="cuda", dtype=torch.float32)
try:
    X1 = torch.linalg.solve(A, b1)
    print("  ✓ SUCCESS")
except Exception as e:
    print(f"  ✗ FAILED: {e}")

# Test case 2: 2D RHS with single column (5, 1) - THIS SHOULD FAIL
print("\nTest 2: 2D RHS (5, 1)")
b2 = torch.rand(A.size(0), 1, device="cuda", dtype=torch.float32)
try:
    X2 = torch.linalg.solve(A, b2)
    print("  ✓ SUCCESS")
except Exception as e:
    print(f"  ✗ FAILED: {e}")

# Test case 3: 2D RHS with multiple columns (5, 3) - should work
print("\nTest 3: 2D RHS (5, 3)")
b3 = torch.rand(A.size(0), 3, device="cuda", dtype=torch.float32)
try:
    X3 = torch.linalg.solve(A, b3)
    print("  ✓ SUCCESS")
except Exception as e:
    print(f"  ✗ FAILED: {e}")
```

Fixes test: test_linalg_solve_sparse_csr_cusolver

zitongzhan committed:
minimal example

zitongzhan committed:
add the numel check for spsolve

zitongzhan committed:
remove unrelated files

zitongzhan committed:
update spsolve's doc

zitongzhan committed:
out no longer needed

zitongzhan committed:
bump up cudss version

pearu reviewed (changes requested):
I have a couple of nits and a request to add samples that test the introduced feature of multiple r.h.s.-s. Otherwise looks good!

zitongzhan committed:
move back `out`

zitongzhan committed:
Apply suggestion from @pearu

Co-authored-by: Pearu Peterson <pearu.peterson@gmail.com>

zitongzhan committed:
fix csr solve in `linalg_solve_out`

zitongzhan committed:
nits

zitongzhan committed:
place back `b.numel() > 0` guard

pearu commented on test/test_sparse_csr.py:
```diff
@@ -3465,22 +3465,18 @@ def test_linalg_solve_sparse_csr_cusolver(self, device, dtype):
-            if sample.args[0].ndim != 1 and sample.args[0].size(-1) != 1:
-                with self.assertRaisesRegex(RuntimeError, "b must be a 1D tensor"):
-                    out = torch.linalg.solve(sample.input.to_sparse_csr(), *sample.args, **sample.kwargs)
-                break
```
Rather than removing the test, please generalize it:
```python
if sample.args[0].ndim not in {1, 2} ...:
    with self.assertRaisesRegex(RuntimeError, "b must be a 1D or 2D tensor"):
        ...
    continue
```

pearu reviewed (changes requested):
Thanks for the update @zitongzhan!

I have a couple of nits and a suggestion to avoid unnecessary tensor copies when possible.

zitongzhan commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -800,10 +811,41 @@ Tensor _sparse_csr_prod_cuda(const Tensor& input, IntArrayRef dims_to_reduce, bo
+  if (b.dim() == 2) {
+    // cuDSS expects column-major dense matrices. Convert the row-major contiguous
+    // RHS to a column-major strided view backed by a contiguous transpose.
+    Tensor b_col_major = b.transpose(-2, -1).contiguous().transpose(-2, -1);
```
Thanks for the suggestion!

Unfortunately, cuDSS currently does not support row-major dense matrices, even though the API exposes the `CUDSS_LAYOUT_ROW_MAJOR` enum. Internally, cuDSS only let the RHS to be column-major. Because of this, we convert `b` to a column-major contiguous layout regardless before calling into cuDSS.

If cuDSS adds proper row-major support in the future, we could revisit this and avoid the extra copy when possible.

https://docs.nvidia.com/cuda/cudss/types.html#c.cudssLayout_t.CUDSS_LAYOUT_ROW_MAJOR:~:text=Row%2Dmajor%20layout.%20Currently%20not%20supported.

zitongzhan committed:
Update aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

Co-authored-by: Pearu Peterson <pearu.peterson@gmail.com>

zitongzhan committed:
fix https://github.com/pytorch/pytorch/pull/171035/changes/BASE..e8a41dfce38501452562669b2857d37ff01d6f11#r2910559988

zitongzhan committed:
Update aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

Co-authored-by: Pearu Peterson <pearu.peterson@gmail.com>

zitongzhan committed:
amend

ZihangFang committed:
restore cudss version number

zitongzhan committed:
constructs a genuine 3D RHS and asserts the raise "b must be a 1D or 2D tensor"

zitongzhan committed:
Test the out=  branches

zitongzhan committed:
update the decomposition expectation
the branch adds/registers aten::_spsolve.out, and HasDecompTest.test_has_decomposition expects every traceable ATen op without a decomposition to be listed in that expect file. aten::_spsolve was already listed; the new out overload was missing.

zitongzhan committed:
fix always always allocates `x_col_major`, then `out.copy_(x_col_major)`

zitongzhan committed:
fix test error "cudss state invalid"

original `Tensor b_col_major = b.transpose(-2, -1).contiguous().transpose(-2, -1);` does not solve the problem

zitongzhan committed:
quality. fix name space consistency

zitongzhan committed:
quality. Redundant guard in the 1D fast path.

zitongzhan committed:
quality. torch/sparse/__init__.py docstring

zitongzhan committed:
quality. "x columns count must be the same as b" reads awkwardly
